File: 2954-maximum-sum-of-almost-unique-subarray/2954-maximum-sum-of-almost-unique-subarray.py
class Solution:
    def maxSum(self, nums: List[int], m: int, k: int) -> int:
        maxSum = 0
        currentSum = 0
        elementFrequency = {}
        left = 0

        for right in range(len(nums)):
            currentElement = nums[right]

            # Update frequency and sum
            elementFrequency[currentElement] = elementFrequency.get(currentElement, 0) + 1
            currentSum += currentElement

            # Shrink the window if it exceeds k elements
            while right - left + 1 > k:
                leftElement = nums[left]

                # Update frequency and sum
                elementFrequency[leftElement] -= 1
                if elementFrequency[leftElement] == 0:
                    del elementFrequency[leftElement]
                currentSum -= leftElement

                # Move the left pointer to shrink the window
                left += 1

            # Check if the subarray has at least m distinct elements
            if len(elementFrequency) >= m:
                maxSum = max(maxSum, currentSum)

        return maxSum

File: 2954-maximum-sum-of-almost-unique-subarray/2954-maximum-sum-of-almost-unique-subarray.js
/**
 * @param {number[]} nums
 * @param {number} m
 * @param {number} k
 * @return {number}
 */
var maxSum = function(nums, m, k) {
    let maxSum = 0;
    let currentSum = 0;
    const elementFrequency = new Map();
    let left = 0;

    for (let right = 0; right < nums.length; right++) {
        const currentElement = nums[right];

        // Update frequency and sum
        elementFrequency.set(currentElement, (elementFrequency.get(currentElement) || 0) + 1);
        currentSum += currentElement;

        // Shrink the window if it exceeds k elements
        while (right - left + 1 > k) {
            const leftElement = nums[left];

            // Update frequency and sum
            elementFrequency.set(leftElement, elementFrequency.get(leftElement) - 1);
            if (elementFrequency.get(leftElement) === 0) {
                elementFrequency.delete(leftElement);
            }
            currentSum -= leftElement;

            // Move the left pointer to shrink the window
            left++;
        }

        // Check if the subarray has at least m distinct elements
        if

 (elementFrequency.size >= m) {
            maxSum = Math.max(maxSum, currentSum);
        }
    }

    return maxSum;
};